Use simple parser instead of regex for parsing resolv.conf

geoand opened the pull request:
- Closes: #5429

tsegismont reviewed (commented):
Thank you @geoand 

Have you got a chance to write a micro-benchmark and compare results?

geoand commented:
Not yet

geoand commented:
Actually I am not sure what a microbenchmark would look like in this case as I've removed the old methods and replaced them with a single one that extracts the necessary data in one pass.

@franz1981 how do you usually handle cases like this? Moreover, do you think a microbenchmark is even necessary in this case?

franz1981 commented:
That's how I do it: https://github.com/franz1981/java-puzzles/blob/74d60bbb8639ccbc1158046ce8d5575720bca96c/src/main/java/red/hat/puzzles/cstatic/StaticInt.java

You can use -prof jfr:dir=profile-results\;configName=jfr-profile.jfc

And use JMH to read the class loading events and compilation events using both -prof cl -prof comp.

These are very tricky benchmarks since:
- class loading can be dominated by I/O files so having enough forks is necessary to make it happen on the OS page cache - but it means it will be less representative of a real use case since the OS page cache would appear like a super fast disk (whilst is not what happen on containers with rate limited disks i.e. AWS)
- beware whatever is loaded by JMH infra which can affect measurements: e.g. if what you want to benchmark is leading an expensive Java  util class A, but the JMH infra does it too, it will make the code under test to appear faster, If it was the dominant factor, but maybe they same class won't be loaded in the real use case

Which means, make sure the effect you want to measure is big enough that observing it, reproduces it with fidelity and is not absorbed by the infrastructure cost

geoand commented:
Here is what a CPU time flamegraph looks like currently:

![orig](https://github.com/user-attachments/assets/18cee79a-67af-4cae-af27-b540caf7e10b)

And with this patched backported to 4.5.11 it looks like this:

![patched](https://github.com/user-attachments/assets/103c4613-3d0e-4471-8da8-e8b37c8cee23)

geoand commented:
It does seem however that this is causing more allocations so I'll see what I can do

geoand commented:
Ideally I would want this in Vert.x 4 as well :)

tsegismont commented:
> Actually I am not sure what a microbenchmark would look like in this case as I've removed the old methods and replaced them with a single one that extracts the necessary data in one pass.

It would have two methods comparing the code inside `if (isLinux()) { ... }` before and after the change.



> I pushed an update and now things look rather nice.
> 
> Here is the CPU time flamegraph:

I'm not an expert but it seems both diagrams use `AddressResolver` init as the wide base. So the conclusion that performance has been improved is drawn from the number of samples ? (i.e 60 instead of 118)

geoand commented:
> So the conclusion that performance has been improved is drawn from the number of samples ? (i.e 60 instead of 118)

Correct.

> I'm not an expert but it seems both diagrams use AddressResolver init as the wide base

I could have also used a base lower down the call stack that would show that `AddressResolver#<clinit>` is now less wide

geoand commented:
Anything more you would like from this PR?

tsegismont commented:
> > Actually I am not sure what a microbenchmark would look like in this case as I've removed the old methods and replaced them with a single one that extracts the necessary data in one pass.
> 
> It would have two methods comparing the code inside `if (isLinux()) { ... }` before and after the change.

Have you been able to write a JMH test comparing the two implementations?

geoand commented:
I have not even tried it. Do you really think it's necessary given the evidence above?

gsmet commented on vertx-core/src/main/java/io/vertx/core/impl/HostnameResolver.java:
```diff
@@ -45,35 +45,26 @@ public class HostnameResolver implements AddressResolver {
-
-  private static final Pattern NDOTS_OPTIONS_PATTERN = resolvOption("ndots:[ \\t\\f]*(\\d)+");
-  private static final Pattern ROTATE_OPTIONS_PATTERN = resolvOption("rotate");
+  private static final Pattern WHITESPACE_PATTERN = Pattern.compile("\\s+");
```
I'm curious if the following would avoid the apparently heavy initialization of what I suspect is the usage of `\s`
```suggestion
  private static final Pattern WHITESPACE_PATTERN = Pattern.compile("[ \\t]+");
```

Asking because I suspect `resolv.conf` would support only traditional spaces and tabs.

Now maybe it's not worth it.

geoand replied:
It's certainly worth a try. I'll be out for a few days, so I can try it when I get back if you haven't already

gsmet replied:
I have no idea which app you used to generate this so I think it can wait for a few days.

geoand replied:
I just used the rest-json Quarkus quickstart on a Linux machine

gsmet replied:
Actually this regexp is not used anymore, AFAICS.

geoand replied:
Updated

gsmet commented on vertx-core/src/main/java/io/vertx/core/impl/HostnameResolver.java:
```diff
@@ -129,18 +120,65 @@ public Future<Void> close() {
+          int ndotsIndex = afterOptions.indexOf(NDOTS_LABEL);
+          if (ndotsIndex != -1) {
+            if (!containsNewLine(afterOptions.substring(0, ndotsIndex))) {
+              Matcher matcher = Pattern.compile("ndots:[ \\t\\f]*(\\d)+(?=$|\\s)").matcher(afterOptions.substring(ndotsIndex));
```
I see you're not compiling this anymore.

I suppose you have good reasons (like it being used only once?).

geoand replied:
It should probably be moved into a holder instead

geoand replied:
Done

gsmet commented:
OK, what's cool now is that we should be entirely regexp-free in the simple case so I would expect the flamegraph to look even better than the one you posted.

geoand commented:
Indeed!

vietj commented on vertx-core/src/main/java/io/vertx/core/impl/HostnameResolver.java:
```diff
@@ -190,4 +227,31 @@ public void dispose(L data) {
+      return ndots;
+    }
+
+    public int getEffectiveNdots() {
```
rename to `effectiveNdots`

vietj commented on vertx-core/src/main/java/io/vertx/core/impl/HostnameResolver.java:
```diff
@@ -190,4 +227,31 @@ public void dispose(L data) {
+    }
+  }
+
+  private static class Holder {
```
is this static class needed because of lazy loading purpose ?

vietj commented on vertx-core/src/main/java/io/vertx/core/DeploymentOptions.java:
```diff
@@ -42,6 +43,7 @@ public class DeploymentOptions {
   private int workerPoolSize;
   private long maxWorkerExecuteTime;
   private TimeUnit maxWorkerExecuteTimeUnit;
+  private String exactDeploymentId;
```
this should be removed from this PR

geoand replied:
Good point! done

tsegismont reviewed (approved):
LGTM, thank you @geoand

geoand commented:
🙏🏽
